src/main/kotlin/com/ampnet/projectservice/service/impl/ProjectServiceImpl.kt

```diff
@@ -80,22 +81,32 @@ class ProjectServiceImpl(
     }
 
     @Transactional
-    override fun updateProject(project: Project, request: ProjectUpdateRequest): Project {
-        validateRoi(request.roi)
-        request.name?.let { project.name = it }
-        request.description?.let { project.description = it }
-        request.location?.let {
-            project.location.lat = it.lat
-            project.location.long = it.long
+    override fun updateProject(serviceRequest: ProjectUpdateServiceRequest): Project {
+        validateRoi(serviceRequest.request.roi)
+        serviceRequest.request.name?.let { serviceRequest.project.name = it }
+        serviceRequest.request.description?.let { serviceRequest.project.description = it }
+        serviceRequest.request.location?.let {
+            serviceRequest.project.location.lat = it.lat
+            serviceRequest.project.location.long = it.long
         }
-        request.roi?.let {
-            project.roi.from = it.from
-            project.roi.to = it.to
+        serviceRequest.request.roi?.let {
+            serviceRequest.project.roi.from = it.from
+            serviceRequest.project.roi.to = it.to
         }
-        request.active?.let { project.active = it }
-        request.tags?.let { it -> project.tags = it.toSet().map { tag -> tag.toLowerCase() } }
-        request.news?.let { project.newsLinks = it }
-        return projectRepository.save(project)
+        serviceRequest.request.active?.let { serviceRequest.project.active = it }
+        serviceRequest.request.tags?.let {
+            it ->
+            serviceRequest.project.tags = it.toSet().map { tag -> tag.toLowerCase() }
+        }
+        serviceRequest.request.news?.let { serviceRequest.project.newsLinks = it }
+        serviceRequest.image?.let {
+            addMainImageToProject(it, serviceRequest.project)
+        }
+        serviceRequest.documentSaveRequests?.forEach {
```

Comment:
Store documents in parallel.
```suggestion
        serviceRequest.documentSaveRequests?.parallelStream()?.forEach {
```